kong/runloop/handler.lua

```diff
@@ -210,6 +210,11 @@ local function balancer_setup_stage1(ctx, scheme, host_type, host, port,
   ctx.route            = route
   ctx.balancer_data    = balancer_data
   ctx.balancer_address = balancer_data -- for plugin backward compatibility
+
+  -- plugin forwards-compatible way
+  kong.ctx.shared.service       = service
+  kong.ctx.shared.balancer_data = balancer_data
```

Comment:
I've seen a WIP commit floating around that changed the name (and possibly content?) of this `balancer_data` structure. Not sure what will happen of that structure, but just a heads up. Also, I don't know what the plugin is doing with it, but we did add functions in `kong/pdk/service.lua` to avoid having plugins work on `balancer_data` directly. (I just noticed we only added setters and not getters — what does the plugin need?)

Reply:
The trick is getting this one merged first then xD

> what does the plugin need?

It seems it gets statistics about each try, the total count, and the ip involved: https://github.com/Kong/kong-plugin-zipkin/blob/master/kong/plugins/zipkin/opentracing.lua#L185-L201

Reply:
I have rolled back the usage of kong.ctx by zipkin - it was a bit premature. So I will also roll back this particular change too.